src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c

```diff
@@ -0,0 +1,405 @@
+          uint32x4_t v89AB_1 = vcombine_u32(vget_low_u32(v89_13), vget_low_u32(vAB_13));
+          uint32x4_t v89AB_3 = vcombine_u32(vget_high_u32(v89_13), vget_high_u32(vAB_13));
+          uint32x4_t vCDEF_0 = vcombine_u32(vget_low_u32(vCD_02), vget_low_u32(vEF_02));
+          uint32x4_t vCDEF_2 = vcombine_u32(vget_high_u32(vCD_02), vget_high_u32(vEF_02));
+          uint32x4_t vCDEF_1 = vcombine_u32(vget_low_u32(vCD_13), vget_low_u32(vEF_13));
+          uint32x4_t vCDEF_3 = vcombine_u32(vget_high_u32(vCD_13), vget_high_u32(vEF_13));
+
+          v0123_0 = xnn_packed2planar(&ksum0123, v0123_0, vmask, veor_mask, neg_zp);
+          v0123_1 = xnn_packed2planar(&ksum0123, v0123_1, vmask, veor_mask, neg_zp);
+          v0123_2 = xnn_packed2planar(&ksum0123, v0123_2, vmask, veor_mask, neg_zp);
+          v0123_3 = xnn_packed2planar(&ksum0123, v0123_3, vmask, veor_mask, neg_zp);
+          v4567_0 = xnn_packed2planar(&ksum4567, v4567_0, vmask, veor_mask, neg_zp);
+          v4567_1 = xnn_packed2planar(&ksum4567, v4567_1, vmask, veor_mask, neg_zp);
+          v4567_2 = xnn_packed2planar(&ksum4567, v4567_2, vmask, veor_mask, neg_zp);
+          v4567_3 = xnn_packed2planar(&ksum4567, v4567_3, vmask, veor_mask, neg_zp);
+          v89AB_0 = xnn_packed2planar(&ksum89AB, v89AB_0, vmask, veor_mask, neg_zp);
+          v89AB_1 = xnn_packed2planar(&ksum89AB, v89AB_1, vmask, veor_mask, neg_zp);
+          v89AB_2 = xnn_packed2planar(&ksum89AB, v89AB_2, vmask, veor_mask, neg_zp);
+          v89AB_3 = xnn_packed2planar(&ksum89AB, v89AB_3, vmask, veor_mask, neg_zp);
+          vCDEF_0 = xnn_packed2planar(&ksumCDEF, vCDEF_0, vmask, veor_mask, neg_zp);
+          vCDEF_1 = xnn_packed2planar(&ksumCDEF, vCDEF_1, vmask, veor_mask, neg_zp);
+          vCDEF_2 = xnn_packed2planar(&ksumCDEF, vCDEF_2, vmask, veor_mask, neg_zp);
+          vCDEF_3 = xnn_packed2planar(&ksumCDEF, vCDEF_3, vmask, veor_mask, neg_zp);
+
+          vst1q_u8(&out[0], v0123_0);
+          vst1q_u8(&out[16], v4567_0);
+          vst1q_u8(&out[32], v89AB_0);
+          vst1q_u8(&out[48], vCDEF_0);
+          vst1q_u8(&out[64], v0123_1);
+          vst1q_u8(&out[80], v4567_1);
+          vst1q_u8(&out[96], v89AB_1);
+          vst1q_u8(&out[112], vCDEF_1);
+          vst1q_u8(&out[128], v0123_2);
+          vst1q_u8(&out[144], v4567_2);
+          vst1q_u8(&out[160], v89AB_2);
+          vst1q_u8(&out[176], vCDEF_2);
+          vst1q_u8(&out[192], v0123_3);
+          vst1q_u8(&out[208], v4567_3);
+          vst1q_u8(&out[224], v89AB_3);
+          vst1q_u8(&out[240], vCDEF_3);
+
+          w0 += 16;
+          w1 += 16;
+          w2 += 16;
+          w3 += 16;
+          w4 += 16;
+          w5 += 16;
+          w6 += 16;
+          w7 += 16;
+          w8 += 16;
+          w9 += 16;
+          w10 += 16;
+          w11 += 16;
+          w12 += 16;
+          w13 += 16;
+          w14 += 16;
+          w15 += 16;
+          out += 256;
+        }
+
+        uint16x8_t bf_scales01234567 = {
```

Comment:
does this result in 16 loads? Is it temporary until the transposed_scales flag?

Reply:
yea it is done in place of transposed scales flag. Transposed scales can give ~10% extra uplift. But it's a lot more working wiring things up, so i just decided to do this as short term. -10% perf isn't that bad compared to how much better compared to scalar both in binary size and perf. We can think of transposed scales later, but the one concern about that is what the best way to multi thread that would be, since scales is read only it wouldn't be that bad, but passing scale stride could be a bit nasty later on.